File: samples/bot-message-reaction/python/bots/messageReactionBot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from botbuilder.core import ActivityHandler, MessageFactory,TurnContext

# This class represents a bot that handles message activities and reactions in Microsoft Teams.
class MessageReactionBot(ActivityHandler):
    def __init__(self, app_id, app_password, activity_log):
        """
        Initialize the bot with app credentials and an activity log for storing activity data.
        :param app_id: The Microsoft app ID for the bot.
        :param app_password: The Microsoft app password for the bot.
        :param activity_log: An object to manage logging of activity IDs and their associated data.
        """
        super().__init__()
        self.app_id = app_id
        self.app_password = app_password
        self._log = activity_log  # Reference to an activity log instance for storing and retrieving activities.
        
    async def on_message_activity(self, turn_context:TurnContext, next_handler=None):
        # Remove recipient mention from the activity
        TurnContext.remove_recipient_mention(turn_context.activity)
        text = turn_context.activity.text
        
        """
        Handle incoming message activities.
        :param turn_context: Provides information about the current turn of the conversation.
        :param next_handler: Optional handler to call after the current method.
        """
        # Respond to the user with an echo of their message and log the activity ID.
        await self.send_message_and_log_activity_id(turn_context, f"echo: {text}")   

    async def on_reactions_added(self, reactions_added, turn_context):
        """
        Handle reactions added to messages.
        :param reactions_added: A list of reactions that were added.
        :param turn_context: Provides information about the current turn of the conversation.
        """
        for reaction in reactions_added:
            # Look up the activity in the log using the replyToId (ID of the message being reacted to).
            activity = await self._log.find(turn_context.activity.reply_to_id)

            if not activity:
                # If the activity was not found, notify the user and log the event.
                await self.send_message_and_log_activity_id(
                    turn_context,
                    f"Activity {turn_context.activity.reply_to_id} not found in the log."
                )
            else:
                # If the activity is found, acknowledge the reaction and include the activity's text.
                await self.send_message_and_log_activity_id(
                    turn_context,
                    f" added '{reaction.type}' regarding '{activity.text}'"
                )

    async def on_reactions_removed(self, reactions_removed, turn_context):
        """
        Handle reactions removed from messages.
        :param reactions_removed: A list of reactions that were removed.
        :param turn_context: Provides information about the current turn of the conversation.
        """
        for reaction in reactions_removed:
            # Look up the activity in the log using the replyToId.
            activity = await self._log.find(turn_context.activity.reply_to_id)
            if not activity:
                # If the activity was not found, notify the user.
                await self.send_message_and_log_activity_id(
                    turn_context, 
                    f"Activity {turn_context.activity.reply_to_id} not found in the log."
                )
            else:
                # If the activity is found, acknowledge the removal of the reaction.
                await self.send_message_and_log_activity_id(
                    turn_context, 
                    f"You removed '{reaction.type}' regarding '{activity.text}'"
                )

    async def send_message_and_log_activity_id(self, turn_context, text):
        """
        Send a message to the user and log the activity ID.
        :param turn_context: Provides information about the current turn of the conversation.
        :param text: The message text to send to the user.
        """
        # Create a text message activity.
        reply_activity = MessageFactory.text(text)
        # Send the activity to the user and capture the response containing the activity ID.
        resource_response = await turn_context.send_activity(reply_activity)
        # Log the activity ID and its associated text in the activity log.
        await self._log.append(resource_response.id, reply_activity)

File: samples/bot-message-reaction/nodejs/bots/messageReactionBot.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const {
    ActivityHandler,
    MessageFactory
} = require('botbuilder');

/**
 * MessageReactionBot handles incoming messages and reactions in Microsoft Teams.
 * @class
 */
class MessageReactionBot extends ActivityHandler {
    /**
     * Creates an instance of MessageReactionBot.
     * @param {ActivityLog} activityLog - The activity log for tracking sent messages.
     */
    constructor(activityLog) {
        super();
        this._log = activityLog;
        this.onMessage(async (context, next) => {
            await this._sendMessageAndLogActivityId(context, `echo: ${ context.activity.text }`);
            await next();
        });
    }

    /**
     * Handles reactions added to messages.
     * @param {Array} reactionsAdded - The reactions added to the message.
     * @param {Object} context - The context of the activity.
     */
    async onReactionsAddedActivity(reactionsAdded, context) {
        for (const reaction of reactionsAdded) {
            const activity = await this._log.find(context.activity.replyToId);
            if (!activity) {
                // If we had sent the message from the error handler we wouldn't have recorded the Activity Id and so we shouldn't expect to see it in the log.
                await this._sendMessageAndLogActivityId(context, `Activity ${ context.activity.replyToId } not found in the log.`);
            } else {
                await this._sendMessageAndLogActivityId(context, ` added '${ reaction.type }' regarding '${ activity.text }'`);
            }
        }
    }

    /**
     * Handles reactions removed from messages.
     * @param {Array} reactionsRemoved - The reactions removed from the message.
     * @param {Object} context - The context of the activity.
     */
    async onReactionsRemovedActivity(reactionsRemoved, context) {
        for (const reaction of reactionsRemoved) {
            // The ReplyToId property of the inbound MessageReaction Activity will correspond to a Message Activity that was previously sent from this bot.
            const activity = await this._log.find(context.activity.replyToId);
            if (!activity) {
                // If we had sent the message from the error handler we wouldn't have recorded the Activity Id and so we shouldn't expect to see it in the log.
                await this._sendMessageAndLogActivityId(context, `Activity ${ context.activity.replyToId } not found in the log.`);
            } else {
                await this._sendMessageAndLogActivityId(context, `You removed '${ reaction.type }' regarding '${ activity.text }'`);
            }
        }
    }

    /**
     * Sends a message and logs the activity ID.
     * @param {Object} context - The context of the activity.
     * @param {string} text - The text of the message to send.
     * @private
     */
    async _sendMessageAndLogActivityId(context, text) {
        const replyActivity = MessageFactory.text(text);
        const resourceResponse = await context.sendActivity(replyActivity);
        await this._log.append(resourceResponse.id, replyActivity);
    }
}

module.exports.MessageReactionBot = MessageReactionBot;
